tests/architecture_prototype/test_graph.py

```diff
@@ -0,0 +1,218 @@
+from rasa.architecture_prototype.graph_fingerprinting import (
+    FingerprintStatus,
+    TrainingCache,
+)
+from rasa.architecture_prototype.graph_components import load_graph_component
+from rasa.core.channels import UserMessage
+from rasa.shared.core.constants import ACTION_LISTEN_NAME
+from rasa.shared.core.events import ActionExecuted, UserUttered
+from rasa.shared.core.trackers import DialogueStateTracker
+from rasa.shared.nlu.constants import INTENT_NAME_KEY
+from tests.architecture_prototype import conftest
+from tests.architecture_prototype.graph_schema import (
+    full_model_train_graph_schema,
+    predict_graph_schema,
+    nlu_train_graph_schema,
+)
+import rasa.nlu.registry
+import rasa.core.registry
+
+
+def test_train_nlu():
+    conftest.clean_directory()
+    graph.fill_defaults(nlu_train_graph_schema)
+    serialized = json.dumps(nlu_train_graph_schema)
+    deserialized = json.loads(serialized)
+    graph.visualise_as_dask_graph(deserialized, "nlu_train_graph.png")
+    trained_components = graph.run_as_dask_graph(
+        deserialized,
+        ["train_classifier", "train_response_selector", "train_synonym_mapper"],
+    )
+    print(trained_components)
+
+
+def test_train_full_model(trained_model: Dict):
+    print(trained_model)
+
+
+def test_train_load_predict(prediction_graph: Dict[Text, Any]):
+    graph.visualise_as_dask_graph(prediction_graph, "predict_graph.png")
+
+    predictions = graph.run_as_dask_graph(predict_graph_schema, ["select_prediction"])
+    for prediction in predictions.values():
+        print(prediction)
+
+    # TODO: should we be saving model_metadata?
+    # TODO: stuff like rasa.utils.train_utils.update_deprecated_loss_type changing wont force re-train?
+    # TODO: include more meta data e.g. rasa version etc.
+
+
+def serialize_graph_schema(graph_schema: Dict[Text, Any]) -> Text:
+    to_serialize = copy.copy(graph_schema)
+    for step_name, step_config in to_serialize.items():
+        component_class = step_config["uses"]
+        step_config["uses"] = component_class.name
+    return json.dumps(to_serialize)
+
+
+def deserialize_graph_schema(serialized_graph_schema: Text) -> Dict[Text, Any]:
+    schema = json.loads(serialized_graph_schema)
+    for step_name, step_config in schema.items():
+        component_class_name = step_config["uses"]
+        try:
+            step_config["uses"] = rasa.nlu.registry.get_component_class(
+                component_class_name
+            )
+        except Exception:
+            try:
+                step_config["uses"] = load_graph_component(component_class_name)
+            except Exception:
+                try:
+                    step_config["uses"] = rasa.core.registry.policy_from_module_path(
+                        component_class_name
+                    )
+                except:
+                    raise ValueError("Unknown component!")
+
+    return schema
+
+
+def test_serialize_graph_schema():
```

Comment:
Maybe we have on test whch does all the training, persisting, loading, predicting etc.?